File: apps/web/src/lib/session_rate_stats.ts
/** Aggregate stats for completed-transfer rates within a session (MiB/s). */

export interface RateStats {
  count: number;
  median: number;
  max: number;
  average: number;
}

export const EMPTY_RATE_STATS: RateStats = { count: 0, median: 0, max: 0, average: 0 };

/** Compute count, median, max and average from a list of per-transfer rates. */
export function computeRateStats(rates: readonly number[]): RateStats {
  const values = rates.filter((rate) => rate > 0);
  if (values.length === 0) {
    return EMPTY_RATE_STATS;
  }
  const sorted = [...values].sort((a, b) => a - b);
  const count = sorted.length;
  const mid = Math.floor(count / 2);
  const median = count % 2 === 0 ? (sorted[mid - 1] + sorted[mid]) / 2 : sorted[mid];
  const max = sorted[count - 1];
  const average = sorted.reduce((sum, value) => sum + value, 0) / count;
  return { count, median, max, average };
}

File: sdks/cpp/session/session_rate_stats.cpp
#include "session_rate_stats.h"

#include <algorithm>
#include <cmath>
#include <numeric>

RateStats compute_rate_stats(const std::vector<double>& rates) {
    std::vector<double> values;
    values.reserve(rates.size());
    for (double rate : rates) {
        if (rate > 0.0) {
            values.push_back(rate);
        }
    }
    if (values.empty()) {
        return {};
    }

    std::sort(values.begin(), values.end());
    const int count = static_cast<int>(values.size());
    const int mid = count / 2;
    const double median = (count % 2 == 0)
        ? (values[static_cast<size_t>(mid - 1)] + values[static_cast<size_t>(mid)]) / 2.0
        : values[static_cast<size_t>(mid)];
    const double max = values.back();
    const double average =
        std::accumulate(values.begin(), values.end(), 0.0) / static_cast<double>(count);
    return {count, median, max, average};
}

void SessionRateTracker::observe_result_mbps(double result_mbps) {
    if (result_mbps > 0.0 && prev_result_mbps_ <= 0.0) {
        rates_.push_back(result_mbps);
    }
    prev_result_mbps_ = result_mbps;
}

RateStats SessionRateTracker::stats() const {
    return compute_rate_stats(rates_);
}
